airflow/jobs.py

```diff
@@ -2300,8 +2301,33 @@ def _per_task_process(task, key, ti, session=None):
                 self.log.debug('Adding %s to not_ready', ti)
                 ti_status.not_ready.add(key)
 
+            pool_name = self.pool
+            if self.pool:
+                pool = session.query(models.Pool) \
+                    .filter(models.Pool.pool == self.pool) \
+                    .first()
+                if pool is not None:
+                    open_slots = pool.open_slots(session=session)
+                else:
+                    open_slots = 0
+                    self.log.warning(
+                        "Tasks using non-existent pool '%s' will not be scheduled",
+                        pool,
+                    )
+            else:
+                open_slots = conf.getint('core', 'non_pooled_backfill_task_slot_count')
+                pool_name = 'not_pooled'
+
             for task in self.dag.topological_sort():
                 for key, ti in list(ti_status.to_run.items()):
+                    if open_slots <= 0:
+                        self.log.debug(
+                            "Not scheduling since there are %s open slots in pool %s",
+                            open_slots, pool_name,
+                        )
+                        break
+                    open_slots -= 1
```

Comment:
I feel that just keeping a count in memory isn't right, and we should check the DB for an accurate count of the pool each loop.

I also don't think that breaking out of the loop is the right behaviour - this would mean that if you hit the pool limit backfilling just stops, rather than waiting for a slot to become free.

Reply:
Backfilling would not stop actually. We will wait here for later retry:
https://github.com/apache/airflow/blob/84a7a9e6cb81139a2b62996acbd8579aff576810/airflow/jobs.py#L2484-L2508

Reply:
Changed to check open slots for each task